lib/modules/datasource/rpm/index.ts

```diff
@@ -0,0 +1,212 @@
+    if (!response.body.startsWith('<?xml')) {
+      logger.debug(
+        { datasource: RpmDatasource.id, url: repomdUrl },
+        'Invalid response format',
+      );
+      throw new Error(
+        `${repomdUrl} is not in XML format. Response body: ${response.body}`,
+      );
+    }
+
+    // parse repomd.xml using XmlDocument
+    const xml = new XmlDocument(response.body);
+
+    const primaryData = xml.childWithAttribute('type', 'primary');
+
+    if (!primaryData) {
+      logger.debug(
+        `No primary data found in ${repomdUrl}, xml contents: ${response.body}`,
+      );
+      throw new Error(`No primary data found in ${repomdUrl}`);
+    }
+
+    const locationElement = primaryData.childNamed('location');
+    if (!locationElement) {
+      throw new Error(`No location element found in ${repomdUrl}`);
+    }
+    const href = locationElement.attr.href;
+    if (!href) {
+      throw new Error(`No href found in ${repomdUrl}`);
+    }
+    // replace trailing "repodata/" from registryUrl, if it exists, with a "/" because href includes "repodata/"
+    const registryUrlWithoutRepodata = registryUrl.replace(
+      /\/repodata\/?$/,
+      '/',
+    );
+    return joinUrlParts(registryUrlWithoutRepodata, href);
+  }
+
+  async getReleasesByPackageName(
+    primaryGzipUrl: string,
+    packageName: string,
+  ): Promise<ReleaseResult | null> {
+    let response: HttpResponse<Buffer>;
+    let decompressedBuffer: Buffer;
+    try {
+      // primaryGzipUrl is a .gz file, need to extract it before parsing
+      response = await this.http.getBuffer(primaryGzipUrl);
+      if (response.body.length === 0) {
+        logger.debug(`Empty response body from getting ${primaryGzipUrl}.`);
+        throw new Error(`Empty response body from getting ${primaryGzipUrl}.`);
+      }
+      // decompress the gzipped file
+      decompressedBuffer = await gunzipAsync(response.body);
+    } catch (err) {
+      logger.debug(
+        `Failed to fetch or decompress ${primaryGzipUrl}: ${
+          err instanceof Error ? err.message : err
+        }`,
+      );
+      throw err;
+    }
+
+    // Use sax streaming parser to handle large XML files efficiently
+    // This allows us to parse the XML file without loading the entire file into memory.
+    const releases: Record<string, Release> = {};
+    let insidePackage = false;
+    let isTargetPackage = false;
+
+    // Create a SAX parser in strict mode
+    const saxParser = sax.createStream(true, {
+      lowercase: true, // normalize tag names to lowercase
+      trim: true,
+    });
+
+    saxParser.on('opentag', (node: sax.Tag) => {
+      if (node.name === 'package' && node.attributes.type === 'rpm') {
+        insidePackage = true;
+        isTargetPackage = false;
+      }
+      if (insidePackage && node.name === 'name') {
+        saxParser.on('text', (text: string) => {
+          if (text.trim() === packageName) {
+            isTargetPackage = true;
+          }
+        });
```

Comment:
you'll attach this handler multiple times. move this out of the event handler or remove it properly on close tag event.